### ISSUE TITLE
Explicit handling of categorical variables in KernelExplainer

### ISSUE THREAD

**Issue by jklaise** [shap/shap#451]
First of all, great work on the library! I've been working on implementing  a wrapper around it to provide SHAP explanations for productionised models but have run into some issues concerning categorical variables and would like to know your thoughts.

I have a concern about the handling of categorical variables in the model-agnostic algorithm of KernelExplainer. The way I see it it is not actually explicitly supported, but there are examples with categorical variables (e.g. https://slundberg.github.io/shap/notebooks/Census%20income%20classification%20with%20scikit-learn.html) that are handled implicitly but can be quite misleading.

My point is that handling categorical variables will depend on the specific encoding and the choice of a background dataset (what does it mean to be non-informative for a categorical variable?). In the example above the categorical variables are encoded as ordinals and the background is the median on the training set. This does not make much sense to me for variables that should not conform to an ordinal encoding (e.g. relationship status, race). The above example has additional issues in that it uses a KNN on a mixed variable dataset and applies scaling to said ordinal encodings of the categorical variables.

I think there should be an explicit disclaimer for using KernelExplainer with categorical variables. If they are ordinally encoded, it probably doesn't make sense to take the mean or the median as the background dataset (unless the ordinal encoding can be argued to be consistent with the meaning of the variable, e.g. education level). This is especially important because by default KernelExplainer calculates the background dataset as the mean across features.

On the topic of one-hot-encoded categorical variables, the median would not work well as it would just pick 0 for all values of a categorical variable which are <50% in the training set. The mean in this case actually works well but with the caveat that the value would be out-of-distribution (real number in [0,1] rather than a discrete value in {0,1}).

With this in mind, there are some options of explicitly handling the non-informativeness of categorical variables, albeit this might require some research:
 * LIME-style sampling from the prior category distribution (this could similarly be applied to numerical variables as well instead of just going with the mean). However, this may require multiple samples to reduce the variance of the value of the "missing" variables.
 * The mode could be a better option in specific cases rather than the mean/median of an ordinally encoded variable.

Apart from ordinal and one-hot there are many other encoding of categorical variables which may require further thought, but I think we should handle the simpler cases first.

**Comment by slundberg** [comment]
First of all, thanks for the careful discussion here! ...and sorry for the slow response. It took more than a moment to reply to, then got buried as I pushed for some deadlines.

I agree that notebook should be updated to use better backgrounds, and some notes added to discuss the potential issues with categorical values. Of course all these issues go away if you pass a large background dataset, but that becomes too slow very quickly. I personally don't think "all zeros" is a bad reference for a one-hot encoded set of variables where none of them are on most of the time. Though the mode is certainly better when one case is dominate.

Would you be interested in updating that notebook to highlight these concerns? Or perhaps sharing another notebook illustrating them for KernelExplainer? I'll leave a todo note here as well so in case you can't get to it, I will know to check back and hopefully get to it myself at some point.

**Comment by jklaise** [comment]
Apologies for the late reply. I'm not quite sure what you mean by "the issues will go away" if the background dataset is large enough. Sampling for categorical variables will still be wrong either if they're ordinally encoded or one-hot encoded. My thinking is we should sample the categorical variables in the raw feature space effectively treating any encodings of categorical variables as part of the model. This would mean relatively simplistic sampling (e.g. uniform over the number of classes if you know it in advance or according to the class frequency in the training data) and it would not be local sampling, but at least it would respect the nature of the categorical variables.

A more advanced use case would be if the user provides the encoding/embedding they used on their categorical variables and then we can sample locally in the transformed space, this would require the encoding to be invertible (e.g. no hashing trick) to retain interpretability downstream.

**Comment by alexcoca** [comment]
Hi @slundberg ,

I guess that for this issue, one could group the categorical variables using `shap.common.DenseData` as discussed in a different post.

An additional trick would be to pass the explainer the data that has not been encoded and ensure that the predictor transforms the data as it pleases before returning the targets for the regression.

**Comment by pramitchoudhary** [comment]
Adding my thoughts on this topic.
In general, i think i agree with @slundberg on the use of "all zero" as the background (baseline/reference) data which is common for DNN use-cases as well E.g. Use of empty string for sentiment classification . But, one could argue that providing domain specific baseline could be helpful in generating explanation that might be less surprising.

However, in the absence of having access to such baseline, using "all zero" as a neutral reference aligns in simulating feature absence by replacing them with `0`. 

For categorical, I have not tried `shap.common.DenseData` yet. Will share my findings as I experiment however the trick suggested by @alexcoca could work. Thanks for the suggestion @alexcoca
Also, not sure if the explainer, by default, should try to encode categorical using one of the technique because ideally we would like the feature space to be same for the base model and the regularized additive model, no?

Summary of Experimentation so far in-case helpful,
```
import pandas as pd
from sklearn.model_selection import train_test_split
from joblib import dump, load
import pickle
import seaborn as sns
import matplotlib.pyplot as plt

df = pd.read_csv(
    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
    header=None)
df.columns = [
    "Age", "WorkClass", "fnlwgt", "Education", "EducationNum",
    "MaritalStatus", "Occupation", "Relationship", "Race", "Gender",
    "CapitalGain", "CapitalLoss", "HoursPerWeek", "NativeCountry", "Income"
]
# df = df.sample(frac=0.01, random_state=1)
train_cols = df.columns[0:-1]
label = df.columns[-1]
X = df[train_cols]
features = X.columns

type_info = [str(_d) for _d in list(X.dtypes)]
feature_info = dict(zip(features, type_info))

y = df[label].apply(lambda x: 0 if x == " <=50K" else 1) #Turning response into 0 and 1

# We have to transform categorical variables to use sklearn models
from sklearn.preprocessing import OneHotEncoder

categorical_mask = X.dtypes == object
# Filter categoricals
categorical_cols = features[categorical_mask].tolist()

ohe = OneHotEncoder(handle_unknown='ignore', sparse = False)

# Transform
ohe.fit(df[categorical_cols])
# Save encoding for future downstream task.
cat_ohe = ohe.transform(df[categorical_cols])
dump(ohe, ./data/models/adult_encoder.joblib")
encoding_path = './data/models/adult_encoder.pkl'
with open(encoding_path, "wb") as f:
    pickle.dump(ohe, f)


ohe_df = pd.DataFrame(cat_ohe, columns = ohe.get_feature_names(input_features = categorical_cols))
# concat with original data and drop original columns and create combined frame
_X = pd.concat([X, ohe_df], axis=1).drop(columns = categorical_cols, axis=1)


seed = 1
X_train, X_test, y_train, y_test = train_test_split(_X, y, test_size=0.20, random_state=seed)

## Build Model
from sklearn.ensemble import RandomForestClassifier
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline

pca = PCA(random_state=42)
rf = RandomForestClassifier(n_estimators=100, n_jobs=-1, random_state=42)


blackbox_model = Pipeline([('pca', pca), ('rf', rf)])
blackbox_model.fit(X_train, y_train)


# Compute global importance
import matplotlib.pyplot as plt

feature_names = X_train.columns
feature_imp = rf.feature_importances_
sorted_idx = feature_imp.argsort()

y_ticks = np.arange(0, len(feature_names))
fig, ax = plt.subplots(figsize = (40, 30))
ax.barh(y_ticks, feature_imp[sorted_idx])

ax.set_yticks(y_ticks)
ax.set_yticklabels(feature_names[sorted_idx])
ax.set_title("Feature Importances")
fig.tight_layout()
plt.show()

import shap
features = X_train.columns.to_list()
X_train_summary = shap.kmeans(X_train, 100)

# Compare different reference baseline
reference_median = X_train.median().values.reshape(1, X_train.shape[1])
reference_kmeans = X_train_summary
reference_zeros = np.zeros(X_train.shape[1]).reshape(1, X_train.shape[1])
reference_mode =  X_train.mode().iloc[0, :].values.reshape(1, X_train.shape[1])
reference_mean = X_train.mean().values.reshape(1, X_train.shape[1])

# Initialize explainer
explainer_randomample = shap.KernelExplainer(blackbox_model.predict_proba, X_train.iloc[0:100, :])
explainer_median = shap.KernelExplainer(blackbox_model.predict_proba, reference_median)
explainer_kmeans = shap.KernelExplainer(blackbox_model.predict_proba, reference_kmeans)
explainer_zeros = shap.KernelExplainer(blackbox_model.predict_proba, reference_zeros)
explainer_mode = shap.KernelExplainer(blackbox_model.predict_proba, reference_mode)
explainer_mean = shap.KernelExplainer(blackbox_model.predict_proba, reference_mean)
explainer_tree_reference = shap.TreeExplainer(rf, data=reference_zeros, feature_perturbation="interventional")
explainer_tree = shap.TreeExplainer(rf)



row = X_train[(X_train['Age'] == 50) & (X_train['fnlwgt'] == 83311)]
np.random.seed(42)
shap_values_mean = explainer_mean.shap_values(row)
shap_values_median = explainer_median.shap_values(row)
shap_values_mode = explainer_mode.shap_values(row)
shap_values_zeros = explainer_zeros.shap_values(row)
shap_values_kmeans = explainer_kmeans.shap_values(row)
shap_values_randomsample = explainer_randomample.shap_values(row)
shap_values_tree_zero_reference = explainer_tree_reference.shap_values(row)
shap_values_tree = explainer_tree.shap_values(row)


# Plot the results for Class 0
shap.bar_plot(shap_values_mean[0][0], row.iloc[0,:], max_display=10)
shap.bar_plot(shap_values_median[0][0], row.iloc[0,:], max_display=10)
shap.bar_plot(shap_values_mode[0][0], row.iloc[0,:], max_display=10)
shap.bar_plot(shap_values_zeros[0][0], row.iloc[0,:], max_display=10)
shap.bar_plot(shap_values_kmeans[0][0], row.iloc[0,:], max_display=10)
shap.bar_plot(shap_values_tree_zero_reference[0][0], row.iloc[0,:], max_display=10)
shap.bar_plot(shap_values_tree[0][0], row.iloc[0,:], max_display=15)

## Compute correlation and heatmap to understand similarity in the SHAP values with different references.
data = {
    'baseline_mean': shap_values_mean[0][0],
    'baseline_median': shap_values_median[0][0],
    'baseline_mode': shap_values_mode[0][0],
    'baseline_zero': shap_values_zeros[0][0],
    'baseline_kmeans': shap_values_kmeans[0][0],
    'baseline_random_sample_100': shap_values_randomsample[0][0],
    'tree_explainer_zero_base': shap_values_tree_zero_reference[0][0],
    'tree_explainer_no_base': shap_values_tree[0][0],
    'global_importance': rf.feature_importances_
    
}
result_summary = pd.DataFrame(data)
sns.heatmap(result_summary.corr(), annot=True, cmap='BrBG');
```
**Global Importance:**
<img width="1000" alt="Screen Shot 2022-01-14 at 12 59 36 AM" src="https://user-images.githubusercontent.com/1318029/149487896-fa726169-80b4-4a75-9fae-0a71665f0e9f.png">

**Local Explanation:**
1. background: Mean
<img width="662" alt="Screen Shot 2022-01-14 at 1 34 47 AM" src="https://user-images.githubusercontent.com/1318029/149493198-bfa770ac-3147-4ff6-a8f4-e8d370debe84.png">

2. background: Median
<img width="630" alt="Screen Shot 2022-01-14 at 1 36 57 AM" src="https://user-images.githubusercontent.com/1318029/149493490-6296e0cc-18cc-40ea-9412-5c6f18435200.png">

3. background: Mode
<img width="639" alt="Screen Shot 2022-01-14 at 1 37 06 AM" src="https://user-images.githubusercontent.com/1318029/149493520-f4ba2956-c222-4102-8df4-34625b306e5e.png">

4. background: Zeros
<img width="589" alt="Screen Shot 2022-01-14 at 1 37 13 AM" src="https://user-images.githubusercontent.com/1318029/149493592-aa0c21ee-cdd4-40ab-b14d-18975a007470.png">

5. background: Kmeans
<img width="615" alt="Screen Shot 2022-01-14 at 1 39 18 AM" src="https://user-images.githubusercontent.com/1318029/149493749-b9a4a19a-ab89-48b9-84d2-9c19d84f8381.png">

6. background: 100 radom sample
<img width="644" alt="Screen Shot 2022-01-14 at 1 40 20 AM" src="https://user-images.githubusercontent.com/1318029/149494042-21454b35-3ada-43e2-a014-cf38fc3e9a67.png">

7. background: TreeExplainer, zero reference
<img width="649" alt="Screen Shot 2022-01-14 at 1 43 36 AM" src="https://user-images.githubusercontent.com/1318029/149494446-c11816b5-e277-4c3a-9151-71eda74c1130.png">

8. background: TreeExplainer, no reference
<img width="627" alt="Screen Shot 2022-01-14 at 1 45 10 AM" src="https://user-images.githubusercontent.com/1318029/149494620-54c919b3-b342-4b8e-9da3-023cd3140268.png">

9. Correlation Heatmap:
<img width="864" alt="Screen Shot 2022-01-14 at 1 53 02 AM" src="https://user-images.githubusercontent.com/1318029/149495784-9d22d8e2-04f6-407b-9bfa-369c7ca3effa.png">


**Questions:**
1. If there are large number of features, `random sampling` is applied which also adds to the variance and in-consistency in result observed across runs? Does it make sense to allow the user to set seed values? 
I tried setting random_state globally but not sure if that’s the right thing to do? What do u guys think?
2. For a local explanation, does it make sense to compute 95%CI bound based on the samples generated.
3. Does it make sense to allow a param `features` as part of KernelSHAP for Incorporating causal knowledge by specifying which features to consider for perturbation? As a way to add constraint in perturbing features which are irrelevant?

Other related issues to the topic of background data and output consistency: https://github.com/slundberg/shap/issues/955
